src/View/Form/AbstractContext.php

```diff
@@ -0,0 +1,53 @@
+<?php
+namespace Cake\View\Form;
+
+use Cake\Network\Request;
+
+/**
+ * Provides common functionality for ContextInterface implementations.
+ */
+abstract class AbstractContext implements ContextInterface
+{
+    /**
+     * The request object.
+     *
+     * @var \Cake\Network\Request
+     */
+    protected $_request;
+
+    /**
+     * The request type of the attached form.
+     *
+     * @var string
+     */
+    protected $_requestType;
+
+    /**
+     * Constructor.
+     *
+     * @param \Cake\Network\Request $request The request object.
+     */
+    protected function __construct(Request $request)
+    {
+        $this->_request = $request;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function setRequestType($requestType)
+    {
+        $this->_requestType = $requestType;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function val($field)
```

Comment:
* The HTTP query is available on any HTTP verb, not just GET.
    * Ref: http://guzzle3.readthedocs.org/http-client/request.html
* The HTTP body should be made available to any HTTP verb, even GET.
    * Ref: http://stackoverflow.com/a/983458 (opposite to what guzzle does)
* IMHO, FormHelper should behave towards how HTTP works in practise.
* RFC 2616 http://ietf.org/rfc/rfc2616
    * https://annevankesteren.nl/2007/10/http-methods
e.g.:
```
OPTIONS
GET
HEAD
POST
PUT
DELETE
TRACE
CONNECT
PATCH https://datatracker.ietf.org/doc/rfc5789/    
```

### Suggestion
(untested)
```php
<?php
public function val($field)
{
    if (in_array($this->_requestType, ['post', 'put', 'delete', 'patch']) {
        if isset(($this->_request->data($field))) {
            return $this->_request->data($field);
        }
        // This returns a query value in case a data value has not been returned
        return $this->_request->query($field);
    }
    // else: 'options', 'get', 'head', 'trace', 'connect', ...
    if (isset($this->_request->query($field))) {
        return $this->_request->query($field);
    }
    // This allows setting a HTTP body even via HTTP GET.
    // Query takes precedence as it is a get based form
    return $this->_request->data($field);
}